### ISSUE TITLE
Paragraph Vector accuracy on ICLR15 paper, not achievable.

### ISSUE THREAD

**Issue by stangelid** [mesnilgr/iclr15#1]
Hi Grégoire,

First of all, thank you for making your scripts available, It is much appreciated :)

I used your provided code to replicate the results reported in the ensemble paper:
http://arxiv.org/abs/1412.5335

I noticed that the reported accuracy of 89.3% using _solely_ the paragraph vectors is not achievable on the test set, and is actually what I'm getting on the validation set. All other scores appear to be correct.

Below is the final output of your code, ran in full, without any modifications:

> ['RNNLM']        valid / test 83.32 / 86.38
> ['PARAGRAPH']    valid / **test** 89.38 / **88.94**
> ['NBSVM']        valid / test 91.02 / 91.87
> 
> ['RNNLM', 'PARAGRAPH']   valid / test 89.86 / 90.32 - weigths = [ 0.48245512  0.51754488]
> ['RNNLM', 'NBSVM']       valid / test 90.76 / 92.13 - weigths = [ 0.47791671  0.52208329]
> ['PARAGRAPH', 'NBSVM']   valid / test 91.46 / 92.21 - weigths = [ 0.49545455  0.50454545]
> 
> ['RNNLM', 'PARAGRAPH', 'NBSVM']          valid / test 91.32 / 92.47 - weigths [ 0.31594115  0.33892007  0.34513878]

As you see, the performance of paragraph vectors on the test set appears to be lower than what's reported on the paper. I do, however, get an accuracy of ~89.3% on the validation set.

I repeated the experiments a few times, to make sure this was not due to a random factor (vector initialization/order of SGD updates from multiple threads etc.), but there was no significant variation on accuracy scores. 

Just for my own sanity, could it be that you're reporting the wrong number on the iclr paper (validation, instead of test acc.)? Thanks a lot :)

Best regards, 
Stefanos

**Comment by mesnilgr** [comment]
Hi Stefanos,

thanks for looking into our paper & code!

I have run the paragraph vectors method several times (~30) in different setups (with and w/o shuffling the data). I suspect the problem comes from there. I report here my results on the test set for the ~30 runs, average +- std (min < max):

w/o shuffling 89.11 +- 0.31 (min: 88.304; max: 89.84)
shuffling 89.28 +- 0.09 (min: 89.11; max: 89.44)

I did that a while ago before publishing our paper. Let me know if you observe the same numbers, otherwise I'll re-run it on my side and make sure it is correct and we'll find where the discrepancy comes from.

Best,
Grégoire

**Comment by stangelid** [comment]
Hi Grégoire,

Thanks for your quick response on this :)

I did 25 runs of the paragraph vector method (12 w/o and 13 with shuffling). 

This is a summary of the results I got on the **test** set:
_w/o shuffling_ 92.67 +- 0.09 (min: 92.53, max: 92.82)
_with shuffling_ 88.73 +- 0.12 (min: 88.53, max: 88.9)

And on the **validation** set:
_w/o shuffling_ 92.21 +- 0.54 (min: 91.04, max: 93.06)
_with shuffling_ 89.13 +- 0.23 (min: 88.56, max: 89.46)

All runs used the exact word2vec settings that appear in your scripts.

The 92.67% accuracy on the test set _w/o shuffling_ matches the result reported on the original Paragraph Vector paper:
http://cs.stanford.edu/~quocle/paragraph_vector.pdf
which was later declared invalid by Tomas Mikolov.

I hope this helps!

Best,
Stefanos

**Comment by mesnilgr** [comment]
Thanks for reporting this. My results are reported on the test set after training on valid+train. Maybe it does make a difference. Have you tried?

**Comment by stangelid** [comment]
If I understand correctly, this is what your paragraph.sh script does already.

The following  creates a 50-50 train/test split of the dataset:

> head sentence_vectors.txt -n 25000 | awk ' ... ' > full-train.txt
> head sentence_vectors.txt -n 50000 | tail -n 25000 | awk ' ... ' > test.txt

which is used to train the logistic and test on the test set:

> ../liblinear-1.96/train -s 0 full-train.txt model.logreg
> ../liblinear-1.96/predict -b 1 test.txt model.logreg out.logreg
> tail -n 25000 out.logreg > ../scores/PARAGRAPH-TEST

The validation set is just a subset of the full train set:

> head -n 10000 full-train.txt > small-train.txt
> head -n 22500 full-train.txt | tail -n 10000 >> small-train.txt
> head -n 12500 full-train.txt | tail -n 2500 > valid.txt
> tail -n 2500 full-train.txt >> valid.txt

So, the model that is used to evaluate performance on the test set was trained on full-train.txt which, in fact, is small-train.txt + valid.txt. Is my understanding correct, or have I missed something?

Best,
Stefanos

**Comment by mesnilgr** [comment]
Just checked on a few runs and I think you are right!
I'll update the paper ASAP.
Thanks for reporting this!

**Comment by stangelid** [comment]
You're welcome! I'm glad I was able to help.
Stefanos

**Comment by mesnilgr** [comment]
The paper on arxiv should be updated tomorrow.
Best,
Grégoire
